lib/core/src/sdk.rs

```diff
@@ -1204,20 +1269,33 @@ impl LiquidSdk {
 
         let user_lockup_amount_sat_without_service_fee =
             receiver_amount_sat + claim_fees_sat + server_fees_sat;
-        let boltz_fee_sat = pair.fees.boltz(user_lockup_amount_sat_without_service_fee);
-        let user_lockup_amount_sat = user_lockup_amount_sat_without_service_fee + boltz_fee_sat;
+
+        // The resulting invoice amount contains the service fee, which is rounded up with ceil()
+        // Therefore, when calculating the user_lockup amount, we must also round it up with ceil()
+        let user_lockup_amount_sat = (user_lockup_amount_sat_without_service_fee as f64 * 100.0
+            / (100.0 - pair.fees.percentage))
+            .ceil() as u64;
+        let boltz_fee_sat = user_lockup_amount_sat - user_lockup_amount_sat_without_service_fee;
         self.validate_user_lockup_amount_for_chain_pair(&pair, user_lockup_amount_sat)?;
-        let lockup_fees_sat = self.estimate_lockup_tx_fee(user_lockup_amount_sat).await?;
+
+        let payer_amount_sat = req.prepare_response.total_fees_sat + receiver_amount_sat;
+
+        let lockup_fees_sat = match payer_amount_sat == balance_sat {
+            true => self.estimate_drain_tx_fee().await?,
+            false => {
```

Comment:
You're right:
- if `payer_amount == balance`, then it's a perfect drain (all good)
- if `payer_amount < balance` with a difference of 1-200 sats, then `estimate_lockup_tx_fee_chain_send` fails with insufficient funds
- if `payer_amount < balance` with a bigger difference (> 200 sat), then `estimate_lockup_tx_fee_chain_send` succeeds

It probably depends on the tx size: when we expect change, we have an extra output. When it's a `drain`, there is no change output. So the tx size is bigger in the first case, which is why the minimum fee (and minimum spendable amount) is higher than in the case of `drain`.

What this means for the user:
- if they choose `drain = true`, it will work
- if they choose `drain = false`, but a `receiver_amount` high enough that the `payer_amount` gets very close to the `balance` (within 1-200 sat), `prepare_pay_onchain` will fail with insufficient funds
- if they choose `drain = false`, but a `receiver_amount` of any value below that critical threshold, it will work

IMO this is reasonable. If the user wants to send everything, they should use `drain`.

What do you think?

Reply:
Finding the drain fee will find the max amount the user can pay, but only as a discrete amount (C). Anything before it (in the interval `(B, C)`) will throw "insufficient funds" error.

```
Amounts user can pay (payer_amount)

   A                B    C
---++++++++++++++++++----+----

A = smallest payer_amount = chain swap min + lockup fee
B = C - 200 sat (approx)
C = balance = biggest payer_amount = max user lockup + drain fee

+ = valid payer_amount
- = invalid payer_amount
```

So we actually have 2 maximums: B (for normal payments) and C (for drain), where any amount between them would fail.

For the current behavior, `pay_onchain` will throw an "insufficient funds" error if user chooses an amount between B and C (excluding B and C). As a way to avoid that, I can find B and include it in the `prepare_pay_onchain` response as a new field. Should I go for that?

Reply:
I don't expect users to understand there are "two" maximum. We should have one max value where all amounts underneath should work.

Reply:
The reason I explained the 2 maximums above was to show there is no "maximum" as for other swaps:
- B is the closest to what you asked, but it's no maximum because there is a higher value (C)
- C is the highest value, but some amounts just below won't work.

This has implications for the UX, because if the UX allows B as maximum and the user choses it, this will result in a 1 sat change output. If there is 1 sat left in the wallet after choosing the "maximum", obviously users can get confused. However sending that amount +1 will also not work. (cc @erdemyerebasmaz ). What will work for sending all funds is explicitly calling drain.

We can however call B a "soft maximum" or find a better name. Key point is this is different than for all other swaps.

I implemented a way to get this "soft maximum" (B) in 60ee1c768021810f72bc64a8ada69d35b638185e . Is this going in the right direction?

If yes, it makes sense to include it in the `fetch_onchain_limits` response. Tricky part here is that finding B depends on an optional claim feeerate, so `fetch_onchain_limits` would get an optional feerate argument, which doesn't seem right. However IMO we cannot treat this feerate as empty, as it is a valid input in `prepare_pay_onchain` so it therefore influences the "soft maximum" that can be sent.

IMO exposing such a maximum would confuse the users more than help. Instead, I think the current behavior is reasonable:
- for UX validation, app can allow sending
  - `PayOnchainAmount::Receiver`: any amount between minimum and total balance
  - `PayOnchainAmount::Drain`: with no amount input from the user, but which internally chooses the maximum possible (C)
- when user confirms selection, app calls `prepare_pay_onchain`
- if user chose a `PayOnchainAmount::Receiver` amount between B and C, `prepare_pay_onchain` will throw an `InsufficientBalance` error
  - user is asked to try again with a different amount


Let's discuss in the call tomorrow.

Reply:
Do you understand why lwk wallet fails in amount less than the drain amount? See my question here: https://github.com/breez/breez-sdk-liquid/pull/464#discussion_r1743614630

Reply:
Addressed in the discussion today (not exposing the max anymore).